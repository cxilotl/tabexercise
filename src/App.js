import React, { useState, useEffect, Fragment } from 'react';
import axios from 'axios';

import cssStyles from './App.module.scss';

const apiKey = '9wur7sdh84azzazdt3ye54k4';

const generateNewsList = newsList => {
  let newsContent;
  if (newsList && Array.isArray(newsList) && newsList.length > 0) {
    newsContent = (
      <>
        {
          newsList.map((newsItem, index) => {
            const newsTitle = newsItem.webTitle;
            const newsDescription = newsItem.fields.trailText || '';
            return (
              <Fragment key={index}>
                <article>
                  <h4>{ newsTitle }</h4>
                  <p>{ newsDescription }</p>
                </article>
              </Fragment>
            );
          })
        }
      </>
    );
  }
  return newsContent;
};

function App() {
  const searchType = 'uk-news';
  // const additionalElements = 'image';
  const additionalFields = 'trailText';
  const [news, setNews] = useState([]);

  useEffect(  () => {
    // const url = `https://content.guardianapis.com/search?q=${searchType}&order-by=newest&show-fields=${additionalFields}&show-elements=${additionalElements}&api-key=${apiKey}`;
    const url = `https://content.guardianapis.com/search?q=${searchType}&order-by=newest&show-fields=${additionalFields}&api-key=${apiKey}`;
    const fetchNews = async () => {
      return axios({
        method: 'get',
        url,
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json;charset=utf-8',
        },
        timeout: 20000,
      });
    };
    fetchNews()
      .then(data => {
        if (data.status === 200) {
          const response = data.data.response;
          if (typeof response.status === 'string' && response.status === 'ok' && Array.isArray(response.results)) {
            setNews(response.results);
          }  else {
            throw new Error('No data returned');
          }
        }
      })
      .catch(err => {
        console.log('ERR', err);
      });

  }, []);
  return (
    <>
      <header>
        <h1>Tabbed Component</h1>
      </header>
      <main className={ cssStyles.app }>
        <h3>UK News</h3>
        <section>
          { generateNewsList(news) }
        </section>
      </main>
    </>
  );
}

export default App;