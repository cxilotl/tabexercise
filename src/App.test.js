import React from 'react';
import renderer from 'react-test-renderer';
import { render, wait } from '@testing-library/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { default as ukNews } from '../fixtures/uk-news';
import App from './App';

test('Application snapshot', () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Given a page with a tabbed component, it', () => {

  it('should render the page title', () => {
    const { getByText } = render(<App />);
    const appTitle = getByText(/Tabbed Component/i);
    expect(appTitle).toBeInTheDocument();
  });

  describe('When displaying a list of UK news items, it', () => {
    let serverMock;
    let newsUrl;

    beforeEach(() => {
      serverMock = new MockAdapter(axios);
      // https://content.guardianapis.com/search?show-fields=trailText&q=uk-news&api-key=9wur7sdh84azzazdt3ye54k4
      newsUrl = /content\.guardianapis\.com\/search\?/i;
      serverMock.onGet(newsUrl).reply(200, ukNews);
    });

    afterEach(() => {
      serverMock.restore();
    });

    it('should include a tab item', () => {
      const { getByText } = render(<App />);
      const tabTitle = getByText(/UK News/i);
      expect(tabTitle).toBeInTheDocument();
    });

    it('should include a tab content',  async () => {
      expect.assertions(1);

      const { container } = render(<App/>);
      await wait(() => {
        const tabContent = container.getElementsByTagName('article');
        expect(tabContent.length).toBeGreaterThan(5);
      });
    });

  });

});
