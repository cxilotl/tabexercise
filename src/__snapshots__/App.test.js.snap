// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application snapshot 1`] = `
Array [
  <header
    className="header"
  >
    <h1>
      Tabbed Component
    </h1>
  </header>,
  <main
    className="main"
  >
    <nav
      role="navigation"
    >
      <ul
        className="tabLayout"
      >
        <li>
          <h3>
            <a
              href="/uk-news"
              onClick={[Function]}
            >
              UK News
            </a>
          </h3>
        </li>
        <li>
          <h3>
            <a
              href="/football"
              onClick={[Function]}
            >
              Football
            </a>
          </h3>
        </li>
        <li>
          <h3>
            <a
              href="/travel"
              onClick={[Function]}
            >
              Travel
            </a>
          </h3>
        </li>
      </ul>
    </nav>
    <section>
      <ol />
    </section>
  </main>,
]
`;
